docs/source/notes/rref.rst

```diff
@@ -0,0 +1,304 @@
+necessary, and ``UserRRef`` does not hold the data. All usage on the owner will
+retrieve the unique ``OwnerRRef`` instance using the globally unique ``RRefId``.
+A ``UserRRef`` will be created when it is used as an argument or return value in
+``rpc`` or ``remote`` invocation, and the owner will be notified according to
+update the reference count. An ``OwnerRRef`` and its data will be deleted when
+there is no ``UserRRef`` instances globally and there are no reference to the
+``OwnerRRef`` on the owner as well.
+
+
+Assumptions
+^^^^^^^^^^^
+
+RRef protocol is design with the following assumptions.
+
+- **Transient Network Failures**: The RRef design aims to handle transient
+  network failures by retrying messages. Node crashes or permanent network
+  partition is beyond the scope. When those incidents occur, the application
+  may take down all workers, revert to the previous checkpoint, and resume
+  training.
+- **Non-idempotent UDFs**: We assume UDFs are not idempotent and therefore
+  cannot be retried. However, internal RRef control messages will be made
+  idempotent and retryable.
+- **Out of Order Message Delivery**: We do not assume message delivery order
+  between any pair of nodes, because both sender and receiver are using multiple
+  threads. There is no guarantee on which message will be processed first.
+
+
+RRef Lifetime
+^^^^^^^^^^^^^
+
+The goal of the protocol is to delete an ``OwnerRRef`` at an appropriate time.
+The right time to delete an ``OwnerRRef`` is when there are no living
+``UserRRef`` instances and Python GC also agrees to delete the ``OwnerRRef``
+instance on the owner. The tricky part is to determine if there are any living
+``UserRRef`` instances.
+
+Design Reasoning
+----------------
+
+A user can get a ``UserRRef`` in three situations:
+
+1) Receiving a ``UserRRef`` from the owner.
+2) Receiving a ``UserRRef`` from another user.
+3) Creating a new ``UserRRef`` owned by another worker.
+
+
+Case 1 is the simplest where the owner initiates the fork, and hence it can
+easily increment the local reference count. The only requirement is that any
+``UserRRef`` must notify the owner upon destruction. Hence, we need the first
+guarantee:
+
+**G1. The owner will be notified when any ``UserRRef`` is deleted.**
+
+As messages might come delayed or out-of-order, we need one more guarantee to
+make sure the delete message is not processed too soon. If A sends a message to
+B that involves an RRef, we call the RRef on A the parent RRef and the RRef on B
+the child RRef.
+
+**G2. Parent RRef will be deleted until the child RRef is confirmed by the
+owner.**
+
+In cases 2 and 3, it is possible that the owner only partially knows the RRef
+fork graph or not even knowing it at all. For example, an RRef could be
+constructed on a user, and before the owner receives any RPC call, the
+creator user might have already shared the RRef with other users, and those
```

Comment:
Nit: it might be helpful to add a code scenario of where this could happen? Just a couple of lines that create an RRef fork tree by sharing RRefs. This might make it easier for someone to follow - it can be done as a follow up task.

Reply:
ignore this, i just read a bit more and saw the code examples